---
title: "eems_tut"
author: "Tom Swan"
date: "08/02/2021"
output: html_document
---

Load libraries
```{r}

library(tidyverse)
library(devtools)
#install_github("dipetkov/eems/plotting/rEEMSplots")

library(rEEMSplots)
#installing packges to make rEEMSplots work
#install.packages(c("Rcpp","RcppEigen","raster","rgeos","sp"))

library(vcfR)
library(LEA)
library(adegenet)
library(RColorBrewer)
library("rworldmap")
library("rworldxtra")
```

Importing vcf into R
```{r}

#Read in the vcf and convert this to "genid" format to name individuals (indNames function) for sNMF individual bar plots.

vcf_tsi_22 <- read.vcfR("tsi_only2018.vcf", verbose = F)
View(vcf_tsi_22)
dim(vcf_tsi_22)

#vcf has read in. Now convert this to genid.

tsi_22_genid <- vcfR2genind(vcf_tsi_22)

#this file is now genid format.
class(tsi_22_genid)

nLoc(tsi_22_genid) # 25,123 loci
nInd(tsi_22_genid) # 264 individuals
```
Make EEMS input file
```{r}
#from https://github.com/ksil91/Ostrea_PopStructure/blob/master/Making_Files/MakingFilesR.ipynb and https://github.com/dipetkov/eems/tree/master/str2diffs

#start with the genotype matrix

Geno <- tsi_22_genid@tab

#Let's convert the matrix to 0-1 labeling. I arbitrarily choose one allele to be the "derived" allele and, for each individual, count how many copies of the derived allele it carries. This is very easy if the tab matrix is of type "codom".

stopifnot(identical(tsi_22_genid@type, 'codom'))

#Since the labeling does not matter for computing differences, I pick the second allele to label as the "derived" allele. That is, I pick all loci whose name ends with .01.

#Geno <- Geno[, str_detect(colnames(Geno), "\\.01$")]

#now run bef2diffs_v1. Reason to run v1. However, the imputation performed by bed2diffs_v2 would not be appropriate if genotypes are not missing at random. Therefore, rather than using bed2diffs_v2, it would be better to clean the data beforehand and to remove SNPs with high missingness, as it is usually done before any analysis of population structure. And then use bed2diffs_v1.

#Here is a function that implements the "pairwise.complete.obs" method, which corresponds to bed2diffs-v1. The straightforward implementation uses a double loop, so would be slow if the sample size is large.

##

bed2diffs_v1 <- function(Geno) {
  nIndiv <- nrow(Geno)
  nSites <- ncol(Geno)
  Diffs <- matrix(0, nIndiv, nIndiv)
  
  for (i in seq(nIndiv - 1)) {
    for (j in seq(i + 1, nIndiv)) {
      x <- Geno[i, ]
      y <- Geno[j, ]
      Diffs[i, j] <- mean((x - y)^2, na.rm = TRUE)
      Diffs[j, i] <- Diffs[i, j]
    }
  }
  Diffs
}

#

bed2diffs_v2 <- function(Geno) {
  nIndiv <- nrow(Geno)
  nSites <- ncol(Geno)
  Miss <- is.na(Geno)
  ## Impute NAs with the column means (= twice the allele frequencies)
  Mean <- matrix(colMeans(Geno, na.rm = TRUE), ## a row of means
                 nrow = nIndiv, ncol = nSites, byrow = TRUE) ## a matrix with nIndiv identical rows of means
  Mean[Miss == 0] <- 0 ## Set the means that correspond to observed genotypes to 0
  Geno[Miss == 1] <- 0 ## Set the missing genotypes to 0 (used to be NA) 
  Geno <- Geno + Mean
  ## Compute similarities
  Sim <- Geno %*% t(Geno) / nSites
  SelfSim <- diag(Sim) ## self-similarities
  vector1s <- rep(1, nIndiv) ## vector of 1s
  ## This chunk generates a `diffs` matrix
  Diffs <- SelfSim %*% t(vector1s) + vector1s %*% t(SelfSim) - 2 * Sim
  Diffs
  }

##

#now let's compute both versions of the dissimilarity matrix and inspect the eigenvalues.
Diffs_v1 <- bed2diffs_v1(Geno)
Diffs_v1 <- round(Diffs_v1, digits = 6)

Diffs_v2 <- bed2diffs_v2(Geno)
Diffs_v2 <- round(Diffs_v2, digits = 6)

#Check that the dissimilarity matrix has one positive eigenvalue and nIndiv-1 negative eigenvalues, as required by a full-rank Euclidean distance matrix.

sort(round(eigen(Diffs_v1)$values, digits = 2))

sort(round(eigen(Diffs_v2)$values, digits = 2))

#it appears that both are identical - because there is no missing data. Let's use version 1. 

#Write files for V1

write.table(Diffs_v1, "tsi2018-eems.diffs", 
            col.names = FALSE, row.names = FALSE, quote = FALSE)

```

```{r}
#import lat and longs for use with EEMS.

tsi_2018_eems_geo <- read.delim("tsi_geo22.txt", header = F)

write.table(tsi_2018_eems_geo, "tsi2018-eems.coord",
            col.names = FALSE, row.names = FALSE, quote = FALSE)

#note that both files are called the same, but have different file types (.diffs, .coord).
#I will place both files in the same directory (on the SSH) to run the eems.
#I do not have a list of habitat boundary points (.outer), but will see if this can run. 


```

Implementing my data with the rEEMSplots package
```{r}
if (!file.exists(eems_results)) {
 stop("Check that the rEEMSplots package is installed without errors.")
}

#providing path to EEMS run (path_eems) and concatenating the results of all the runs. 

#this path works
path_eems <- "C:/Users/jc486631/OneDrive - James Cook University/Melbourne/Melbourne/Analysis/2021/eems/TSI-EEMS/results/"
dirs_200deems <- c(paste0(path_eems, "tsi2018-200demes-eems-chain1/"), paste0(path_eems, "tsi2018-200demes-eems-chain2/"), paste0(path_eems, "tsi2018-200demes-eems-chain3/"))


#path and dirs attempt v2
```
Plotting the TSI data 200 deme data
```{r}
eems.plots(mcmcpath = dirs_200deems, plotpath = paste0(path_eems, "200deme-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)
```

Plotting the TSI data 250 deme data (200,000 runs, 100,000 burn-in)

```{r}
path_eems <- "C:/Users/jc486631/OneDrive - James Cook University/Melbourne/Melbourne/Analysis/2021/eems/TSI-EEMS/results/"
dirs_250deems <- c(paste0(path_eems, "tsi2018-250demes-eems-chain1/"), paste0(path_eems, "tsi2018-250demes-eems-chain2/"), paste0(path_eems, "tsi2018-250demes-eems-chain3/"))
```

Plotting the 250 deme (5000,000 runs, 1000,000 burn-in))
```{r}
path_eems <- "C:/Users/jc486631/OneDrive - James Cook University/Melbourne/Melbourne/Analysis/2021/eems/TSI-EEMS/results/"
dirs_250deems_5mill <- c(paste0(path_eems, "tsi2018-250demes5millMCMC-eems-chain1/"), paste0(path_eems, "tsi2018-250demes5millMCMC-eems-chain2/"), paste0(path_eems, "tsi2018-250demes5millMCMC-eems-chain3/"))


```

```{r}

eems.plots(mcmcpath = dirs_250deems_5mill, plotpath = paste0(path_eems, "250deme-5mill-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)

eems.plots(mcmcpath = dirs_250deems_5mill, plotpath = paste0(path_eems, "250deme-5mill-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = F,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)
```

250 deme has the best R2 value

```{r}
#now to plot the diversity in colour. 

eems.plots(mcmcpath = dirs_250deems_5mill, plotpath = paste0(path_eems, "250deme-5mill-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T, eems.colors = brewer.pal(11, "RdBu"))



```

Plotting the TSI data 300 deme data (1 mill runs, 100,000 burn-in)

```{r}

#this path works
path_eems <- "C:/Users/jc486631/OneDrive - James Cook University/Melbourne/Melbourne/Analysis/2021/eems/TSI-EEMS/results/"
dirs_300deems <- c(paste0(path_eems, "tsi2018-300demes-eems-chain1/"), paste0(path_eems, "tsi2018-300demes-eems-chain2/"), paste0(path_eems, "tsi2018-300demes-eems-chain3/"))


```

```{r}

eems.plots(mcmcpath = dirs_300deems, plotpath = paste0(path_eems, "300deme-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)

```

Plotting the TSI data 1000 deme data (200,000 runs)

```{r}

#this path works
path_eems <- "C:/Users/jc486631/OneDrive - James Cook University/Melbourne/Melbourne/Analysis/2021/eems/TSI-EEMS/results/"
dirs_1000deems <- c(paste0(path_eems, "tsi2018-1000demes-eems-chain1/"), paste0(path_eems, "tsi2018-1000demes-eems-chain2/"), paste0(path_eems, "tsi2018-1000demes-eems-chain3/"))


eems.plots(mcmcpath = dirs_1000deems, plotpath = paste0(path_eems, "1000deme-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)
```

PLotting the diversity rates in a different colour
```{r}
eems.plots(mcmcpath = dirs_1000deems, plotpath = paste0(path_eems, "1000deme-All-chain-plots"), longlat = T,
           add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T, eems.colors = brewer.pal(11, "RdBu"))


```


Plotting the TSI data 1000 deme data (1 mill runs)

```{r}

dirs_1000deems_1mill <- c(paste0(path_eems, "tsi2018-1000demes-1mill-eems-chain1/"), paste0(path_eems, "tsi2018-1000demes-1mill-eems-chain2/"), paste0(path_eems, "tsi2018-1000demes-1mill-eems-chain3/"))

eems.plots(mcmcpath = dirs_1000deems_1mill, plotpath = paste0(path_eems, "1millMCMC-1000deme-chain-plots"), 
           longlat = T, add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T)

#plotting diversity in a different colour:

eems.plots(mcmcpath = dirs_1000deems_1mill, plotpath = paste0(path_eems, "col-1millMCMC-1000deme-chain-plots"), 
           longlat = T, add.grid=F,add.outline = T, add.demes = T,
           projection.in = "+proj=longlat +datum=WGS84", projection.out = "+proj=merc +datum=WGS84",
           add.map = T,add.abline = T, add.r.squared = T, eems.colors = brewer.pal(11, "RdBu"))

```

Yi Tao's package 
```{r}
#setwd("/path~to~your~folder")

Genetic_diff<-"tsi2018-eems.diffs"
Sample_location<-"tsi2018-eems.coord"
Outline<-"tsi2018-eems.outer"

#calculate the distance, you may need to replace this with the topological skeleton distance
length1<-Edgelength(Genetic_diff, Sample_location, Automode)

#this is the decision graph and MDS 

#let's say you have length value=3600
length1<-10000

#calculate the deme number
deme<-DemeNumber(Outline,length1)

```

```{r}
session_info()

```
